File: src/questions/Pre-Questions/141.LinkedListCycle.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function hasCycle(head: ListNode | null): boolean {
  const nodeSet = new Set();
  nodeSet.add(head?.val);
  let current = new ListNode(0, head);
  let bool = false;

  while (bool) {
    if (current.next === null) break;
    if (nodeSet.has(current.next.val)) {
      bool = true;
    } else {
      nodeSet.add(current.next.val);
      current = current.next;
    }
  }
  return bool;
}

File: src/questions/75-Questions/141.LinkedListCycle.java
import java.util.HashSet;
import java.util.Set;

class ListNode {
  int val;
  ListNode next;

  ListNode(int x) {
    val = x;
    next = null;
  }
}

class Solution {
  public boolean hasCycle(ListNode head) {
    Set<ListNode> set = new HashSet<>();

    ListNode slow = head;
    ListNode fast = head;

    while (fast != null && fast.next != null) {
      slow = slow.next;
      set.add(slow);
      fast = fast.next.next;
      if (set.contains(fast))
        return true;
      set.add(fast);

    }
    return false;
  }
}

// Another solution
/**
 * Definition for singly-linked list.
 * class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 */
// class Solution {
// public boolean hasCycle(ListNode head) {

// ListNode slow = head;
// ListNode fast = head;

// while (fast != null && fast.next != null) {
// slow = slow.next;
// fast = fast.next.next;
// if (slow == fast) return true;

// }
// return false;
// }
// }